timemachine/cpp/src/wrap_kernels.cpp

```diff
@@ -180,6 +180,61 @@ void declare_context(py::module &m) {
         ctxt.get_x_t(buffer.mutable_data());
         return buffer;
     })
+    .def("multiple_steps_delta_U", [](timemachine::Context &ctxt,
+        const double lambda,
+        const int n_steps,
+        const py::array_t<double, py::array::c_style> &lambda_windows,
+        int store_u_interval,
+        int store_x_interval) -> py::tuple {
+        // (ytz): Collect energies across multiple snapshots.
+
+        // Let lambda_windows have shape [K].
+        // The return'd U matrix has F rows and K states. Where F == ceil(n_steps/store_u_interval).
+        // We are guaranteed that for any row U[i, :], the energies are correct up to a constant.
+        // Terms that depend on lambda will always be computed. For example, we may choose to omit
+        // the bonded energies that are independent of lambda from this calculation.
+
+        // Currently however, the energies just so happen to be exact, but this is not
+        // strictly guaranteed. So you should avoid comparing energies down the columns U[:, i].
```

Comment:
These usage comments are very important -- probably will save some missteps if these are made more visible to the Python layer. Would it be okay to switch these from line comments (visible only when inspecting `wrap_kernels.cpp` source) to a docstring (visible interactively via `help(ctxt.multiple_steps_delta_U)`)?

I think this can be done by moving these comments into a string, and then passing that string as the third argument to `.def` https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx

Reply:
It's not clear to me if

1) We need to depend on sphinx explicitly in the build step actually generate the docstrings
2) How help() can look up the generated docstrings
3) What we'd really like is something that pylint can read into (as opposed to something that's only usable interactively). I tested the approach above and it doesn't seem to trigger.

PS I don't mind adding sphinx etc. but we should actually verify that it would actually does what we want to do, namely 3).

Reply:
Ahh, I misunderstood -- will have to look further into how to supply docstrings to pybind.

Re: 1, 2 -- Sorry, didn't mean to suggest adding sphinx dependency, just this was the first search result I encountered when looking for a page in the pybind documentation that shows how to supply a docstring. Supplying a string as the third argument to `.def` and having it accessible via `help(...)` does appear to work in other parts of the pybind documentation e.g. https://pybind11.readthedocs.io/en/stable/basics.html#keyword-arguments , and we had [previously](https://github.com/proteneer/timemachine/pull/441) added annotations to be visible to `help(...)`

Re: 3 -- Hadn't considered pylint compatibility, but that would be helpful for whichever documentation convention we adopt here.

Perhaps "documentation of Python bindings" should be moved into a new issue to avoid derailing the current PR?

Main point I wanted to ensure is visible to the Python layer is that there are per-row offsets. Some applications could be sensitive or insensitive to these offsets, so documenting this somehow to the Python user seems important.

An alternative to a docstring might be to add a keyword argument with a descriptive name for this offset behavior, and have it only accept `True` for now.

Reply:
On second thought, can this offset behavior be simplified / removed altogether?

We want to return `delta_U(x, target_lam) = U(x, target_lam) - U(x, reference_lam)`.

The comment suggests we are instead possibly returning `offset_delta_U(x, target_lam) = delta_U(x, target_lam) + offset(x)`.

Could we instead return `offset_delta_U(x, target_lam) - offset_delta_U(x, reference_lam)` -- which simplifies to `delta_U(x, target_lam)` -- so there is no longer an unspecified offset?

Reply:
> Perhaps "documentation of Python bindings" should be moved into a new issue to avoid derailing the current PR?

Ok - let's settle then on simply having the docstring be accessible via help() for now, and look for a more robust solution (and editor friendly) one in a separate issue/PR.

> Some applications could be sensitive or insensitive to these offsets.

Applications that depend on the total energy probably aren't free energy methods :)

> We want to return `delta_U(x, target_lam) = U(x, target_lam) - U(x, reference_lam)`.

I see that this is bad naming, this should probably be renamed to "offset_U" for clarity.

> `The comment suggests we are instead possibly returning offset_delta_U(x, target_lam) = delta_U(x, target_lam) + offset(x).`

This is correct, we're really just returning energies shifted by a constant amount across the rows.

> Could we instead return offset_delta_U(x, target_lam) - offset_delta_U(x, reference_lam)

Wouldn't returning `offset_delta_U(x, target_lam) - offset_delta_U(x, reference_lam)` require us to return a NxN matrix of energies for *each* frame x against all `target`, `reference` pair of lamba window combinations. Are you suggesting that we explicitly fix the `reference` lambda window  to that of the window used to run the MD simulation (`lambda` argument as opposed to `lambda_windows`)? Note that the reference lambda window *currently* isn't guaranteed to be an element of the "lambda_windows" we evaluate the energies at, but we can change this.

Reply:
Not sure if we can pick an arbitrary `reference_lambda` here... since we want the delta_Us returned to be the same independent of the choice of `reference_lambda` (at least if they were to be used in MBAR/inf-switch).

`U` evaluated is clearly dependent on `lambda`: `U(x,lamb) = U_independent(x) + U_dependent(x, lamb)`. If we were to evaluate at two different values of `lambda`:

`U(x, lamb_ref) = U_independent(x) + U_dependent(x, lamb_ref)`
`U(x, lamb_target) = U_independent(x) + U_dependent(x, lamb_target)`
`delta_U = U_dependent(x, lamb_target) - U_dependent(x, lamb_ref)`

The exact value of the delta_Us we get back out would then be dependent on `lamb_ref`. The offsetting approach only shifts the `U_independent` component by a fixed constant, and is actually what we want.

Reply:
>Wouldn't returning offset_delta_U(x, target_lam) - offset_delta_U(x, reference_lam) require us to return a NxN matrix of energies for each frame x against all target, reference pair of lamba window combinations. 

No -- sorry for unclarity. There is one `reference_lam` (in the `lambda` argument) and multiple `target_lam`s (in the `lambdas` argument).

>Are you suggesting that we explicitly fix the reference lambda window to that of the window used to run the MD simulation (lambda argument as opposed to lambda_windows)?

Yes!

This wouldn't cost any generality, unless I'm missing something? This would just replace an unspecified behavior (how much is each row offset by) with a specified behavior (each row is offset by 0).

>I see that this is bad naming, this should probably be renamed to "offset_U" for clarity.

That might be better!

Summary of improvement options:
* Change definition to return `delta_U`
* Change name or add other user-facing hint that `offset_U` is returned

Reply:
Resolved in https://github.com/proteneer/timemachine/pull/475/commits/7c07f1846a8033ba8b354eb606cb6dbe309639ac by changing function name from `multiple_steps_delta_U` to `multiple_steps_U` and expanding comments into a docstring